### PR TITLE
proc_macro: Stay on the "use the cache" path more

### PR DESCRIPTION
**Pull request by alexcrichton**
Discovered in #50061 we're falling off the "happy path" of using a stringified
token stream more often than we should. This was due to the fact that a
user-written token like `0xf` is equality-different from the stringified token
of `15` (despite being semantically equivalent).

This patch updates the call to `eq_unspanned` with an even more awful solution,
`probably_equal_for_proc_macro`, which ignores the value of each token and
basically only compares the structure of the token stream, assuming that the AST
doesn't change just one token at a time.

While this is a step towards fixing #50061 there is still one regression
from #49154 which needs to be fixed.

### CODE REVIEW

**Comment by nrc** [comment]
Oh my.

@bors: r+ (but I hope you feel bad about this :-) )

**Comment by alexcrichton** [comment]
I do feel bad :(

This is indeed casting a shadow of a doubt on macros 1.2...
